src/async_fuse/memfs/cache/backend.rs

```diff
@@ -0,0 +1,493 @@
+
+        let operator = match config {
+            StorageParams::S3(StorageS3Config {
+                ref endpoint_url,
+                ref access_key_id,
+                ref secret_access_key,
+                ref bucket_name,
+            }) => {
+                let mut builder = S3::default();
+
+                builder
+                    .endpoint(endpoint_url)
+                    .access_key_id(access_key_id)
+                    .secret_access_key(secret_access_key)
+                    .region("auto")
+                    .bucket(bucket_name);
+
+                Operator::new(builder)?.finish()
+            }
+            StorageParams::Fs(ref root) => {
+                let root = if root.is_empty() {
+                    warn!("Argument `--storage-fs-root` is not set while `--storage-type` is set to `fs`. using `/tmp/datenlord_backend` as default.");
+                    "/tmp/datenlord_backend"
+                } else {
+                    root.as_str()
+                };
+
+                let mut builder = Fs::default();
+                builder.root(root);
+                Operator::new(builder)?.finish()
+            }
+        };
+
+        Ok(BackendWrapper {
+            operator,
+            block_size,
+        })
+    }
+}
+
+/// The backend wrapper of `openDAL` operator.
+#[derive(Debug)]
+pub struct BackendWrapper {
+    /// The inner `Operator`
+    operator: Operator,
+    /// Block size
+    block_size: usize,
+}
+
+impl BackendWrapper {
+    /// Create a new backend
+    #[must_use]
+    pub fn new(operator: Operator, block_size: usize) -> Self {
+        Self {
+            operator,
+            block_size,
+        }
+    }
+
+    /// Get file path by `ino`
+    fn get_file_path(ino: INum) -> String {
+        format!("{ino}/")
+    }
+
+    /// Get block path by `ino` and `block_id`
+    fn get_block_path(ino: INum, block_id: usize) -> String {
+        format!("{ino}/{block_id}.block")
+    }
+}
+
+#[async_trait]
+impl Storage for BackendWrapper {
+    async fn load_from_self(&self, ino: INum, block_id: usize) -> Option<Block> {
+        let mut block = Block::new_zeroed(self.block_size);
+
+        if let Err(e) = self
+            .operator
+            .reader(&Self::get_block_path(ino, block_id))
+            .await
+            .unwrap_or_else(|e| {
+                panic!("Failed to get a reader where ino={ino} and block={block_id}: {e}")
+            })
+            .read(block.make_mut_slice())
+            .await
+        {
+            if e.kind() == std::io::ErrorKind::NotFound {
+                None
+            } else {
+                panic!(
+                    "Failed to load a block from backend where ino={ino} and block={block_id}: {e}"
+                );
+            }
+        } else {
+            Some(block)
+        }
+    }
+
+    async fn load_from_backend(&self, _: INum, _: usize) -> Option<Block> {
+        // This storage has no backend.
```

Comment:
Why this function returns None, while the `cache_block_from_backend` just panics? Can we make them consistent?

Reply:
`load_from_backend` will always be called when `load_from_self` returns `None`; and `cache_block_from_backend` will never be called if `load_from_backend` always returns `None`. This is decided by the default implementation of `load()` method.

I can remove the `unreachable!` in `cache_block_from_backend`, but I think panic in a function that will never be called is more semantically clear.

Reply:
I think a panic is better here to prevent accidental operations.

Reply:
> I think a panic is better here to prevent accidental operations.

But I can't. This method will be called by the default implementation of `load()`, when `load_from_self` returns `None`. That means, a calling to this method is a normal situation, while the calling to `cache_block_from_backend` is not. So I can't raise a panic in this method.